iroh-p2p/src/node.rs

```diff
@@ -1095,7 +1266,290 @@ mod tests {
+            })) => {
+                assert_eq!(test_runner_b.peer_id, peer_id);
+                assert_eq!(topic, subscribed_topic);
+            }
+            Some(n) => {
+                anyhow::bail!("unexpected network event: {:?}", n);
+            }
+            None => {
+                anyhow::bail!("expected NetworkEvent::Gossipsub(Subscribed), received no event");
+            }
+        };
+
+        let peers = test_runner_a.client.gossipsub_all_peers().await?;
+        assert!(peers.len() == 1);
+        let got_peer = peers.get(0).unwrap();
+        assert_eq!(test_runner_b.peer_id, got_peer.0);
+        assert_eq!(&topic, got_peer.1.get(0).unwrap());
+
+        // get mesh peer for topic
+        let peers = test_runner_a
+            .client
+            .gossipsub_mesh_peers(topic.clone())
+            .await?;
+
+        assert!(peers.len() == 1);
+        assert_eq!(test_runner_b.peer_id, *peers.get(0).unwrap());
+
+        let peers = test_runner_a.client.gossipsub_all_mesh_peers().await?;
+        assert!(peers.len() == 1);
+        assert_eq!(&test_runner_b.peer_id, peers.get(0).unwrap());
+
+        let msg = Bytes::from(&b"hello world!"[..]);
+        test_runner_b
+            .client
+            .gossipsub_publish(topic.clone(), msg.clone())
+            .await?;
+
+        match test_runner_a.network_events.recv().await {
+            Some(NetworkEvent::Gossipsub(GossipsubEvent::Message { from, message, .. })) => {
+                assert_eq!(test_runner_b.peer_id, from);
+                assert_eq!(topic, message.topic);
+                assert_eq!(test_runner_b.peer_id, message.source.unwrap());
+                assert_eq!(msg.to_vec(), message.data);
+            }
+            Some(n) => {
+                anyhow::bail!("unexpected network event: {:?}", n);
+            }
+            None => {
+                anyhow::bail!("expected NetworkEvent::Gossipsub(Message), received no event");
+            }
+        };
+
+        test_runner_b
+            .client
+            .gossipsub_unsubscribe(topic.clone())
+            .await?;
+        match test_runner_a.network_events.recv().await {
+            Some(NetworkEvent::Gossipsub(GossipsubEvent::Unsubscribed {
+                peer_id,
+                topic: unsubscribe_topic,
+            })) => {
+                assert_eq!(test_runner_b.peer_id, peer_id);
+                assert_eq!(topic, unsubscribe_topic);
+            }
+            Some(n) => {
+                anyhow::bail!("unexpected network event: {:?}", n);
+            }
+            None => {
+                anyhow::bail!("expected NetworkEvent::Gossipsub(Unsubscribed), received no event");
+            }
+        };
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_dht() -> Result<()> {
+        // set up three nodes
+        // two connect to one
+        // try to connect via id
+        let cid: Cid = "bafkreieq5jui4j25lacwomsqgjeswwl3y5zcdrresptwgmfylxo2depppq"
```

Comment:
not a code comment but rather me learning ipfs...
I assume this CID refers to some block or file or something stored on IPFS somewhere.  How does this code ever get that cid?  And what does start_providing and stop_providing do with it?

Reply:
So the CID "Content Identifier" refers to some block of content. When you provide content on the dht, the `kad` behaviour refers to the identifiers as `Key`s rather than `CID`s, so I'm assuming that means you can provide things other than block content over the dht, though I've never wired it up that way.

`start_providing` makes the CID "available" on the dht. Basically, you store it in the `RecordStore`, and you periodically send messages to the peers you have that are "closest" to the key saying "hey I have this". These records have timeouts, so they are only valid for a certain period of time. I believe the TTL defaults to 24hrs. (https://docs.rs/libp2p/latest/libp2p/kad/struct.Kademlia.html#method.start_providing for more details)

`stop_providing` removes the record from your store. This happens locally & doesn't push any sort of "stop providing" records, it just won't re-provide the record during the next interval. (https://docs.rs/libp2p/latest/libp2p/kad/struct.Kademlia.html#method.stop_providing)

As for how we obtain those CIDs, we currently only call `start_providing` after we have created a DAG from some content that was added to iroh. We create the DAG and call `start_providing` with the root hash of the DAG. This happen in `iroh` after we call `add`

We don't call `stop_providing` anywhere.

The kad record store is currently a mem store, so these records get cleared every time you re-start your node. We have plans to re-provide any root cids from the `iroh-store` on start-up for 0.2.0, but currently each time you restart your node you would need to "re-add" the content from the command line to trigger any providing. Details here: https://github.com/n0-computer/iroh/issues/353